text/0000-pub-restricted.md

```diff
@@ -0,0 +1,924 @@
+   easier to read it too.
+
+Developers may misuse this form and make it hard to access the tasty
+innards of other modules.
+
+ * This is true, but I claim it is irrelevant.
+
+   The effect of this change is solely on the visibility of items
+   *within* a crate. No rules for inter-crate access change.
+
+   From the perspective of cross-crate development, this RFC changes
+   nothing, except that it may lead some crate authors to make some
+   things no longer universally `pub` that they were forced to make
+   visible before due to earlier limitations. I claim that in such
+   cases, those crate authors probably always intended for such items
+   to be non-`pub`, but language limitations were forcing their hand.
+
+   As for intra-crate access: My expectation is that an individual
+   crate will be made by a team of developers who can work out what
+   mutual visibility they want and how it should evolve over time.
+   This feature may affect their work flow to some degree, but they
+   can choose to either use it or not, based on their own internal
+   policies.
+
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Do not extend the language!
+
+ * Change privacy rules and make privacy analysis "smarter"
+   (e.g. global reachabiliy analysis)
+
+   The main problem with this approach is that we tried it, and it
+   did not work well: The implementation was buggy, and the user-visible
+   error messages were hard to understand.
+
+   See discussion when the team was discussing the [public items amendment][]
+
+[public items amendment]: https://github.com/rust-lang/meeting-minutes/blob/master/weekly-meetings/2014-09-16.md#rfc-public-items
+
+ * "Fix" the mental model of privacy (if necessary) without extending
+   the language.
+
+   The alternative is bascially saying: "Our existing system is fine; all
+   of the problems with it are due to bugs in the implementation"
+
+   I am sympathetic to this response. However, I think it doesn't
+   quite hold up. Some users want to be able to define items that are
+   exposed outside of their module but still restrict the scope of
+   where they can be referenced, as discussed in the [motivation][]
+   section, and I do not think the current model can be "fixed" to
+   support that use case, at least not without adding some sort of
+   global reachability analysis as discussed in the previous bullet.
+
+In addition, these two alternatives do not address the main point
+being made in the [motivation][] section: one cannot tell exactly how
+"public" a `pub` item is, without working backwards through the module
+tree for all of its re-exports.
+
+## Curb your ambitions!
+
+ * Instead of adding support for restricting to arbitrary module
+   subtrees, narrow the feature to just `pub(crate) item`, so that one
+   chooses either "module private" (by adding no modifier), or
+   "universally visible" (by adding `pub`), or "visible to just the
+   current crate" (by adding `pub(crate)`).
```

Comment:
> I think it is ill-advised to make crates more significant than they already are - we would end up with two layers of modularisation.

I agree that this is not ideal, but how is this not already the case? Coherence rules operate at the crate rather than module level, `extern crate` declarations and cargo dependencies are in themselves not transparent, etc. There is already a pretty big difference in the relation between modules inside my crate and relations with modules outside of it.